### ISSUE TITLE
Binance WebSocket suddenly fails to connect

### ISSUE THREAD

**Issue by LiquidSwapFinance** [knowm/XChange#3912]
Long time XChange user here. Two days ago, connecting to Binance Streaming stopped working.

The error is as follows: 
```
INFO  - [Connecting to wss://stream.binance.com:9443/stream?streams=nanobtc@depth/ethbtc@depth/xrpusdt@depth/btcusdt@depth/trxusdt@depth/linkusdt@depth/dogeusdt@depth/ethusdt@depth/neobtc@depth/ltcusdt@depth] - at info.bitrich.xchangestream.binance.BinanceStreamingService.lambda$openConnection$3(NettyStreamingService.java:144)
ERROR - [WebSocket client [id: 0xac9738da, L:/192.168.1.23:49730 - R:stream.binance.com/18.182.253.31:9443] encountered exception (CodecException - invalid WebSocket Extension handshake for "permessage-deflate; server_no_context_takeover; client_max_window_bits=15"). Closing] - at info.bitrich.xchangestream.service.netty.WebSocketClientHandler.exceptionCaught(WebSocketClientHandler.java:117)
WARN  - [Problem with connection                                                         ] - at info.bitrich.xchangestream.binance.BinanceStreamingService.lambda$openConnection$4(NettyStreamingService.java:269)
io.netty.handler.codec.CodecException: invalid WebSocket Extension handshake for "permessage-deflate; server_no_context_takeover; client_max_window_bits=15"
	at io.netty.handler.codec.http.websocketx.extensions.WebSocketClientExtensionHandler.channelRead(WebSocketClientExtensionHandler.java:110)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:286)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:436)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:324)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:296)
	at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:251)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1533)
	at io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1282)
	at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1329)
	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:508)
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:447)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:276)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:719)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:655)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:581)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:832)
ERROR - [Reconnection failed: invalid WebSocket Extension handshake for "permessage-deflate; server_no_context_takeover; client_max_window_bits=15"] - at info.bitrich.xchangestream.binance.BinanceStreamingService.lambda$scheduleReconnect$7(NettyStreamingService.java:294)
INFO  - [WebSocket Client disconnected! [id: 0xac9738da, L:/192.168.1.23:49730 ! R:stream.binance.com/18.182.253.31:9443]] - at info.bitrich.xchangestream.service.netty.WebSocketClientHandler.channelInactive(WebSocketClientHandler.java:55)
INFO  - [Reopening Websocket Client because it was closed! [id: 0xac9738da, L:/192.168.1.23:49730 ! R:stream.binance.com/18.182.253.31:9443]] - at info.bitrich.xchangestream.binance.BinanceStreamingService.channelInactive(NettyStreamingService.java:529)
INFO  - [Scheduling reconnection                                                         ] - at info.bitrich.xchangestream.binance.BinanceStreamingService.scheduleReconnect(NettyStreamingService.java:285)
```

The code in question is the following:

 ```

ProductSubscription.ProductSubscriptionBuilder builder = ProductSubscription.create();
currencyPairs.forEach(currencyPair -> {
   builder.addOrderbook(currencyPair);
});

StreamingExchange streamingExchange = StreamingExchangeFactory.INSTANCE.createExchange(BinanceStreamingExchange.class);
var spec = streamingExchange.getDefaultExchangeSpecification();
spec.setExchangeSpecificParametersItem(info.bitrich.xchangestream.util.Events.BEFORE_API_CALL_HANDLER, (Runnable) rateLimiter::acquire);
final var streamingExchangeFinal = StreamingExchangeFactory.INSTANCE.createExchange(spec);
streamingExchangeFinal.connect(builder.build()).blockingAwait();
```

And it's on that blockingAwait that the error repeatedly happens, retrying every 10 seconds or so, only to get the same error.

This code hasn't changed in months and has been working fine, so I'm not sure what changed to start causing this. Any help would be appreciated! :)

**Comment by 0xfaulty** [comment]
already fixed in [#3908](https://github.com/knowm/XChange/pull/3908)

**Comment by earce** [comment]
Fixed in earlier review as stated
